### DISCUSSION TITLE
Use unasync approach

### DISCUSSION THREAD

**Discussion by karpetrosyan** [encode/httpx#3046] [Ideas]
We have over 800 lines of code in our `Client` class. 
We also have an `AsyncClient`, which is the same `Client` but with async/await syntax, and we should literally write the same code twice.

We can consider including the same `unasync.py` file from httpcore](https://github.com/encode/httpcore/blob/master/unasync.py) in our CI to ensure that `Client` and `AsyncClient` are always in sync.

It's also more robust and secure because issues like #3023 and #3042 are eliminated.

**Comment by rafalkrupinski** [comment]
Would it be an option to implement the Client around generators, which works with both sync and async, and then make Client and AsyncClient facades?

**Reply by karpetrosyan** [reply]
Sorry for delayed answer, I think unasync is much easier and more familiar to the maintainers. Also, i don't know if the generators would work for the tests.

**Reply by rafalkrupinski** [reply]
I was just hinting to a potential solution, it's really a matter of preference, generated code vs architecture. Httpx is already using generators for auth flows.
